### DISCUSSION TITLE
The difference between Flex and this work

### DISCUSSION THREAD

**Discussion by lyd405121** [InteractiveComputerGraphics/PositionBasedDynamics#107] [General]
I have learned a lot from this  repository and your paper, and I want to implement this repository fully on gpu .
Recent days,  I am learning the Flex engine and Matthias Müller's《Ten Minute Physics》
Here's my video about my work:
![ik](https://user-images.githubusercontent.com/16344694/143983952-8a418238-a8fe-4c17-8eec-7c0c3ff0ec7f.gif)

Here is the difference  I found in rigidbody simulation between this repository and Flex:

In this repository :
model files(ply) → aabb(or kdtree) → collision detect → conact found → sovle constraint

And in Flex:
model files(ply) → particle(discretization, like sdf) → invisible(because Flex is not open source project)

So, the rigidboy simuliton in this repository don't include any particles,  just vertex from the model files. Is that right?
I guess you and Matthias Müller, Miles Macklin are old friends, maybe you know well about the Flex engine   :)

**Comment by korzen** [comment]
The [Unified Particle Physics for Real-Time Applications](https://mmacklin.com/uppfrta_preprint.pdf) paper explains how Flex works internally

**Reply by lyd405121** [reply]
Thanks very very much!
I have glanced over this before,  and unified solver is a great idea, which is convenient for the developer.
Get it， I will dive into it again！ ^_^

**Comment by janbender** [comment]
In this paper they explain that they sample the rigid bodies with particles and then use shape matching to simulate the rigid bodies. In contrast our library uses a proper rigid body model as described (for example) here: 
https://animation.rwth-aachen.de/publication/0513/

**Reply by lyd405121** [reply]
Yes, this paper I read about half of it, too difficult for me now.
So, Can I say the rigidbody simulaiton in this  repository is more precise than Flex?
I found the rigidbody will deform as  dragging it fast in Flex, this maybe the drawback of shape match:
![flex-shape-match](https://user-images.githubusercontent.com/16344694/144039077-7d8d359d-9271-4a37-95aa-53db2981445f.gif)

Thanks for your response， I should pluck up the courage to read it again!

**Reply by janbender** [reply]
That looks funny. A rigid body should not deform at all. Using a proper rigid body model you a rigid body can not deform by definition. It is defined by its mass. inertia tensor, position, rotation, linear and angular velocity. So it only has 6 degrees of freedom: It can be translated in x,y,z-direction and rotated around the 3 axes.

**Reply by lyd405121** [reply]
I got it. 
Ready to study more about it.
So even your paper looks difficult with lots of equations,  but the implement of the solver is not that complicated. 
I think the key problem is how to move all those data structure to gpu unlike Flex,  just particles

**Reply by korzen** [reply]
Have a look also at [this recent paper from NVidia](https://matthias-research.github.io/pages/publications/PBDBodies.pdf)

**Reply by lyd405121** [reply]
> Have a look also at [this recent paper from NVidia](https://matthias-research.github.io/pages/publications/PBDBodies.pdf)

xpbd for ik solve, downloading!

**Reply by zhanglepy** [reply]
> 知道了。准备对此进行更多研究。因此，即使您的论文在有很多方程的情况下看起来也很困难，但是求解器的实现并不那么复杂。我认为关键问题是如何将所有这些数据结构移动到GPU上，而不像Flex，只是粒子

Hello, I would like to ask two questions, is it possible to do deformation simulation of tumors in soft tissue in FleX? Is it possible to get information on the positional changes of each particle that makes up the tumor?

**Comment by lyd405121** [comment]
## The answer 
- Flex using a morph system even for rigid body
- So if dragging an object too fast, the morph animation will not act in time
- But the rigid body matrix is right

## Pictures
- Some pictures made recently after reading your papers

- Rigidbody Simulation

![rigidbody](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/db47e8a9-e905-41c2-be21-63d9704be02c)

- Friction Static and Dynamic

![friction](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/74cc4740-e2c4-4c97-a613-426bbc464475)


- Cloth Simulation with multilayer

![cloth](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/5d93b6ef-7842-41b1-92ab-0f84e4b802bd)


- Cloth Simulation with animated mesh

![dance](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/73b45d51-ee24-4dfa-9909-29a637e60581)


- Soft body Simulation

![softbody](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/3e472d9e-6bdb-4fc8-b771-45d9163fe01a)

## PBD in the wild
- We are using PBD to simulate softbody couple with rigidbody to produce synthetic data for robot grasping task

- The raytraced picture, wire part using softbody, the head part using rigidbody

![000000](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/1b6f15c2-2643-4364-8b43-f62df472e846)

- The annotation automatic and accutate by code

![anno](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/assets/16344694/f51b6039-2f07-441d-81cf-15242f667fc9)
